catalogd/internal/storage/localdir.go

```diff
@@ -20,95 +26,266 @@ import (
-		_ = file.Close()
-		return nil, os.ErrNotExist
+
+	f, err := os.Create(catalogIndexFilePath(catalogDir))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	enc := json.NewEncoder(f)
+	enc.SetEscapeHTML(false)
+	return enc.Encode(idx)
+}
+
+func (s *LocalDirV1) BaseURL(catalog string) string {
+	return s.RootURL.JoinPath(catalog).String()
+}
+
+func (s *LocalDirV1) StorageServerHandler() http.Handler {
+	mux := http.NewServeMux()
+
+	mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "all").Path, s.handleV1All)
+	if s.EnableQueryHandler {
+		mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "query").Path, s.handleV1Query)
+	}
+	return mux
+}
+
+func (s *LocalDirV1) handleV1All(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	serveJsonLines(w, r, catalogStat.ModTime(), catalogFile)
+}
+
+func (s *LocalDirV1) handleV1Query(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	defer catalogFile.Close()
+
+	schema := r.URL.Query().Get("schema")
+	pkg := r.URL.Query().Get("package")
+	name := r.URL.Query().Get("name")
+
+	if schema == "" && pkg == "" && name == "" {
+		// If no parameters are provided, return the entire catalog (this is the same as /api/v1/all)
+		serveJsonLines(w, r, catalogStat.ModTime(), catalogFile)
+		return
+	}
+	idx, err := s.getIndex(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	indexReader, ok := idx.Get(catalogFile, schema, pkg, name)
+	if !ok {
+		httpError(w, fs.ErrNotExist)
+		return
+	}
+	serveJsonLines(w, r, catalogStat.ModTime(), indexReader)
+}
+
+func (s *LocalDirV1) catalogData(catalog string) (*os.File, os.FileInfo, error) {
+	catalogFile, err := os.Open(catalogFilePath(s.catalogDir(catalog)))
+	if err != nil {
+		return nil, nil, err
+	}
+	catalogFileStat, err := catalogFile.Stat()
+	if err != nil {
+		return nil, nil, err
+	}
+	return catalogFile, catalogFileStat, nil
+}
+
+func httpError(w http.ResponseWriter, err error) {
+	var code int
+	switch {
+	case errors.Is(err, fs.ErrNotExist):
+		code = http.StatusNotFound
+	case errors.Is(err, fs.ErrPermission):
+		code = http.StatusForbidden
+	case errors.Is(err, ErrInvalidParams):
+		code = http.StatusBadRequest
+	default:
+		code = http.StatusInternalServerError
+	}
+	http.Error(w, fmt.Sprintf("%d %s", code, http.StatusText(code)), code)
+}
+
+func serveJsonLines(w http.ResponseWriter, r *http.Request, modTime time.Time, rs io.ReadSeeker) {
+	w.Header().Add("Content-Type", "application/jsonl")
+	http.ServeContent(w, r, "", modTime, rs)
+}
+
+func (s *LocalDirV1) getIndex(catalog string) (*index, error) {
+	idx, err, _ := s.sf.Do(catalog, func() (interface{}, error) {
```

Comment:
Let's add some comments here about the purpose of the single flight structure.

Reply:
The TL;DR is that we expect `getIndex` to be called concurrently if there are concurrent HTTP requests being handled by the query handler. The single flight instance gives us a way to load the index from disk exactly once per concurrent group of callers, and then let every concurrent caller have access to the loaded index. This avoids lots of unnecessary open/decode/close cycles when concurrent requests are being handled, which improves overall performance and decreases response latency.

Reply:
Added the comment in the `LocalDirV1` structure definition